### PR TITLE
numpyro 0.15.3 :snowflake: 

### PR DESCRIPTION
**Pull request by Andrii-Osipov**
numpyro 0.15.3 :snowflake:

**Destination channel:** Snowflake

- [PKG-5831](https://anaconda.atlassian.net/browse/PKG-5831) 
- [Upstream repository](https://github.com/pyro-ppl/numpyro)
- [PyPi](https://pypi.org/project/numpyro/)
- [conda-forge](https://github.com/conda-forge/numpyro-feedstock)

[PKG-5831]: https://anaconda.atlassian.net/browse/PKG-5831?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,34 +1,44 @@
...
+  # https://github.com/jax-ml/jax/blob/jax-v0.4.35/setup.py#L118
+  # Missing jax and jaxlib for s390x
+  skip: True  # [py<=39 or py>311 or s390x]
 
 requirements:
   host:
-    - python >=3.6
+    - python
     - pip
+    - wheel
```
Comment (lorepirri) [inline recipe/meta.yaml:25]:
```suggestion
    - wheel
    - setuptools
```

Comment (lorepirri) [reply recipe/meta.yaml:25]: https://github.com/pyro-ppl/numpyro/blob/3cde93d0f25490b9b90c1c423816c6cfd9ea23ed/setup.py#L9

File: recipe/meta.yaml [hunk 2]
```diff
@@ -1,34 +1,44 @@
...
 
 build:
-  noarch: python
+  script: {{ PYTHON }} -m pip install . -vv --no-deps --no-build-isolation
   number: 0
-  script: {{ PYTHON }} -m pip install .
+  # Missing jax and jaxlib python version 3.9
+  # https://github.com/jax-ml/jax/blob/jax-v0.4.35/setup.py#L118
+  # Missing jax and jaxlib for s390x
+  skip: True  # [py<=39 or py>311 or s390x]
```
Comment (lorepirri) [inline recipe/meta.yaml:19]:
```suggestion
  skip: True  # [py<=39 or s390x]
```

`jax/jaxlib` are available for `py312`

File: recipe/meta.yaml [hunk 3]
```diff
@@ -1,34 +1,44 @@
...
-    - python >=3.6
+    - python
     - pip
+    - wheel
   run:
-    - jax >=0.2
-    - jaxlib >=0.1.55
+    - python
+    - jax >=0.4.25 # [py>39]
+    - jaxlib >=0.4.25 # [py>39]
```
Comment (lorepirri) [inline recipe/meta.yaml]:
```suggestion
    - jax >=0.4.25
    - jaxlib >=0.4.25
```
why the pinnings?

File: recipe/meta.yaml [hunk 4]
```diff
@@ -1,34 +1,44 @@
...
     - multipledispatch
     - numpy
-    - python >=3.6
     - tqdm
 
+# Skipped pytest by dependencies packages missing reason.
```
Comment (lorepirri) [inline recipe/meta.yaml]: could you please run the tests that can be run without `pyro-api` and `funsor`?

**Review by lorepirri — COMMENTED** [review]
a few things

**Comment by lorepirri** [comment]
https://github.com/pyro-ppl/numpyro/blob/0.15.3

File: recipe/meta.yaml (hunk 4 above)
Comment (Andrii-Osipov) [reply recipe/meta.yaml]: Screenshot from 2024-11-14 13-04-22

File: recipe/meta.yaml (hunk 2 above)
Comment (Andrii-Osipov) [reply recipe/meta.yaml:19]: But numpyro does not support 3.12.
https://github.com/pyro-ppl/numpyro/blob/master/setup.py#L102

File: recipe/meta.yaml (hunk 3 above)
Comment (Andrii-Osipov) [reply recipe/meta.yaml]: @lorepirri Comments added
Jax v0.4.35 is not support python 3.9
https://github.com/jax-ml/jax/blob/jax-v0.4.35/setup.py#L118
If we need to add compatibility with python3.9 we must add Jax v0.4.25
https://github.com/jax-ml/jax/blob/jax-v0.4.25/setup.py#L188
That is why I skipped 3.9

File: recipe/meta.yaml (hunk 2 above)
Comment (lorepirri) [reply recipe/meta.yaml:19]: that is metadata, the actual constrain is this one https://github.com/pyro-ppl/numpyro/blob/0.15.3/setup.py#L87

Comment (lorepirri) [reply recipe/meta.yaml:19]: if it does not build, then we skip py312 with a comment

File: recipe/meta.yaml (hunk 4 above)
Comment (lorepirri) [reply recipe/meta.yaml]: I will DM you

File: recipe/meta.yaml (hunk 3 above)
Comment (lorepirri) [reply recipe/meta.yaml]: this is unfortunate, if `jax` is not optional, then this package will need `jax` always in the run section. We sadly need to skip `py39` completely.

File: recipe/meta.yaml (hunk 4 above)
Comment (lorepirri) [reply recipe/meta.yaml]: I meant: skip the tests that require those two dependencies.

File: recipe/meta.yaml [hunk 5]
```diff
@@ -38,12 +38,16 @@ requirements:
 # Skipped pytest by dependencies packages missing reason. 
 # pyro-api >=0.1.1 and funsor >=0.4.1
 test:
+  source_files:
+    - test
   imports:
     - numpyro
   commands:
     - pip check
+    - pytest -v --ignore=test/test_pickle.py test/test_handlers.py
```
Comment (lorepirri) [inline recipe/meta.yaml]: In this way you are telling `pytest` to run the tests that are only in the file `test/test_handlers.py`. Is this the purpose?

File: recipe/meta.yaml [hunk 6]
```diff
@@ -1,34 +1,117 @@
...
 
 package:
-  name: numpyro
+  name: {{ name|lower }}
   version: {{ version }}
 
 source:
-  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
-  sha256: d357dd1a0159f86d45c233708ad33a3d3d53431d05b4bbece1d068ae4168aa8e
+  #url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
```
Comment (skupr-anaconda) [inline recipe/meta.yaml]:
```suggestion
```
It's redundant

File: recipe/meta.yaml [hunk 7]
```diff
@@ -1,34 +1,117 @@
...
 
 build:
-  noarch: python
+  script: {{ PYTHON }} -m pip install . -vv --no-deps --no-build-isolation
   number: 0
-  script: {{ PYTHON }} -m pip install .
+  # Missing jax and jaxlib python version 3.9
+  # https://github.com/jax-ml/jax/blob/jax-v0.4.35/setup.py#L118
+  # Missing jax and jaxlib for s390x
+  # The numpyro max version python 3.11
```
Comment (skupr-anaconda) [inline recipe/meta.yaml:18]: Will it work with py312? Or do some of the dependencies have constraints?

File: recipe/meta.yaml [hunk 8]
```diff
@@ -1,34 +1,117 @@
...
+    - pip check
+    - pytest -v {{ ignore_modules }} {{ deselect_tests }} test
+  requires:
+    - pip
+    - pytest
+    # optax is not avaliable on win-64 and osx-64 
+    # osx-arm64 consist optax but don`t have openssl[version='>=1.1.1j,<1.1.2a as dep by optax
+    - optax>=0.0.6 # [not (win or osx)]
+    - scikit-learn
+    - matplotlib
```
Comment (skupr-anaconda) [inline recipe/meta.yaml]:
```suggestion
    - matplotlib-base
```
A lightweight variant without `pyqt`

**Review by skupr-anaconda — APPROVED** [review]
LGTM but minor questions

File: recipe/meta.yaml (hunk 7 above)
Comment (Andrii-Osipov) [reply recipe/meta.yaml:18]: Only 3.9-3.11

Comment (ViridianMelody) [reply recipe/meta.yaml:18]: I see 3.9-3.11 tagged in `setup.py`, but the actual constraint is `python_requires=">=3.9"`. Might be worth at least attempting to build for 3.12 unless there's some specific dependency constraint.

File: recipe/meta.yaml (hunk 2 above)
Comment (lorepirri) [reply recipe/meta.yaml:19]: @Andrii-Osipov

File: recipe/meta.yaml [hunk 9]
```diff
@@ -1,34 +1,116 @@
...
     - numpyro
+  commands:
+    - pip check
+    - pytest -v {{ ignore_modules }} {{ deselect_tests }} test
+  requires:
+    - pip
+    - pytest
+    # optax is not avaliable on win-64 and osx-64 
+    # osx-arm64 consist optax but don`t have openssl[version='>=1.1.1j,<1.1.2a as dep by optax
+    - optax>=0.0.6 # [not (win or osx)]
```
Comment (lorepirri) [inline recipe/meta.yaml]:
```suggestion
    - optax >=0.0.6 # [not (win or osx)]
```

**Review by ViridianMelody — APPROVED** [review]
LGTM

File: recipe/meta.yaml (hunk 7 above)
Comment (lorepirri) [reply recipe/meta.yaml:18]: @Andrii-Osipov can we try to build for py312 as well before?

**Pull request merged**